In [184]:
import numpy as np
import pandas as pd

In [2]:
from collections import Counter

In [3]:
import spacy

In [4]:
nlp = spacy.load("en_core_web_sm")

In [5]:
from datasets import load_dataset

dataset = load_dataset("taln-ls2n/inspec")

/Users/nishanthkrishna/miniforge3/envs/nlp/lib/python3.10/site-packages/datasets/load.py:1461: FutureWarning: The repository for taln-ls2n/inspec contains custom code which must be executed to correctly load the dataset. You can inspect the repository content at https://hf.co/datasets/taln-ls2n/inspec
You can avoid this message in future by passing the argument `trust_remote_code=True`.
Passing `trust_remote_code=True` will be mandatory to load this dataset from the next major release of `datasets`.
  warnings.warn(


In [6]:
df = pd.DataFrame(dataset['train'])
test_df = pd.DataFrame(dataset['test'])
val_df = pd.DataFrame(dataset['validation'])

In [7]:
def get_pos(my_text):
    nlp_tokens = nlp(my_text)
    return [token.pos_ for token in nlp_tokens]

def get_tokens(my_text):
    nlp_tokens = nlp(my_text)
    return [token for token in nlp_tokens]

def get_key_pos(my_list):
    res = []
    for keyword in my_list:
        phrase = []
        for token in nlp(keyword):
            phrase.append(token.pos_)
        res.append(phrase)
    return res

In [8]:
df['tokens'] = df['abstract'].apply(get_tokens)
df['pos'] = df['abstract'].apply(get_pos)
df['key_pos'] = df['keyphrases'].apply(get_key_pos)

In [9]:
df

id                                              title  \
0     761  Towards a NMR implementation of a quantum latt...   
1     724  Banking on SMA funds [separately managed accou...   
2    1371  Design methodology for diagnostic strategies f...   
3    1334                          A shy invariant of graphs   
4    1419  PacketVideo. One step ahead of the streaming w...   
..    ...                                                ...   
995  1124  Data extraction from the Web based on pre-defi...   
996   118       Sensorless control of induction motor drives   
997   635  Detection and estimation of abrupt changes in ...   
998  1260  A dataflow computer which accelerates executio...   
999  1225                       BT voices its support for IP   

                                              abstract  \
0    Recent theoretical results suggest that an arr...   
1    From investment management to technology to ba...   
2    This paper presents a method for the construct...   
3    Moving from a well known result of P.L. Hammer...   
4    Go beyond the hype, however, and it's clear th...   
..                                                 ...   
995  With the development of the Internet, the Worl...   
996  Controlled induction motor drives without mech...   
997  Detection of change-points in normal means is ...   
998  In the dataflow machine, it is important to av...   
999  BTexact's chief technology officer, Mick Reeve...   

                                            keyphrases  \
0    [NMR implementation, quantum lattice gas algor...   
1    [separately managed accounts, investment manag...   
2    [design methodology, diagnostic strategies, in...   
3    [graph invariant, induced odd cycles, minimum ...   
4    [PacketVideo, wireless devices, MPEG-4, wirele...   
..                                                 ...   
995  [data extraction, schema, Internet, informatio...   
996  [sensorless control, induction motor drives, r...   
997  [generalized likelihood ratio test statistic, ...   
998  [dataflow computer, sequential programs, prece...   
999  [BTexact, voice over DSL, virtual private netw...   

                                                  prmu  \
0                             [P, P, P, P, P, P, P, M]   
1                                   [P, P, P, P, P, P]   
2                                      [P, P, P, P, P]   
3                                   [P, P, P, P, P, P]   
4                             [P, P, P, P, P, P, P, R]   
..                                                 ...   
995                        [P, P, P, P, P, R, M, U, U]   
996  [P, P, P, P, P, P, P, P, P, P, P, P, P, P, P, ...   
997  [P, P, P, P, P, P, P, P, P, P, P, P, P, P, P, ...   
998               [P, P, P, P, P, P, P, P, P, R, M, M]   
999                                       [P, P, P, P]   

                                                tokens  \
0    [Recent, theoretical, results, suggest, that, ...   
1    [From, investment, management, to, technology,...   
2    [This, paper, presents, a, method, for, the, c...   
3    [Moving, from, a, well, known, result, of, P.L...   
4    [Go, beyond, the, hype, ,, however, ,, and, it...   
..                                                 ...   
995  [With, the, development, of, the, Internet, ,,...   
996  [Controlled, induction, motor, drives, without...   
997  [Detection, of, change, -, points, in, normal,...   
998  [In, the, dataflow, machine, ,, it, is, import...   
999  [BTexact, 's, chief, technology, officer, ,, M...   

                                                   pos  \
0    [ADJ, ADJ, NOUN, VERB, SCONJ, DET, NOUN, ADP, ...   
1    [ADP, NOUN, NOUN, ADP, NOUN, ADP, ADJ, PUNCT, ...   
2    [DET, NOUN, VERB, DET, NOUN, ADP, DET, NOUN, A...   
3    [VERB, ADP, DET, ADV, VERB, NOUN, ADP, PROPN, ...   
4    [VERB, ADP, DET, NOUN, PUNCT, ADV, PUNCT, CCON...   
..                                                 ...   
995  [ADP, DET, NOUN, ADP, DET, NOUN, PUNCT,

In [10]:
df['keyphrases'][10]

['SCADA',
 'supervisory control',
 'data acquisition',
 'in-house integration',
 'compatibility',
 'programmable logic controllers']

In [11]:
df['key_pos'][10]

[['PROPN'],
 ['ADJ', 'NOUN'],
 ['NOUN', 'NOUN'],
 ['ADP', 'PUNCT', 'NOUN', 'NOUN'],
 ['NOUN'],
 ['ADJ', 'NOUN', 'NOUN']]

In [12]:
count = 0
for i in df['key_pos']:
    count+=len(i)
print(count)

9805


In [113]:
def temp(a_list):
    sl = [a_list[0]]
    # print(f"added element 0 - {a_list[0]}")
    i=1
    while((i>0) and (i<len(a_list))):
        # print(f"i is at {i} - {a_list[i]}")
        if a_list[i]==a_list[i-1] and (i<len(a_list)):
            # print("same as last term")
            while ((i<len(a_list)) and (a_list[i]==a_list[i-1])):
                # print(f"incrementing i from {i}")
                i+=1
                # print(f"to {i}")
            # if i<len(a_list):
                # print(f"SAME: i is at {i} - {a_list[i]}")
            # print(f"last one is {sl[-1]}")
            sl[-1] += "+"
        if (i<len(a_list)) and a_list[i]!=a_list[i-1]:
            # print(f"added element i - {a_list[i]}")
            sl.append(a_list[i])
            i+=1
    # print(",".join(sl))
    return ",".join(sl)

In [114]:
temp(['ADP', 'PUNCT', 'NOUN', 'NOUN', 'NOUN', 'NOUN', 'PUNCT'])

'ADP,PUNCT,NOUN+,PUNCT'

In [115]:
for i in df['key_pos'][10]:
    print(i)
    temp(i)
    print("---"*15)

['PROPN']
---------------------------------------------
['ADJ', 'NOUN']
---------------------------------------------
['NOUN', 'NOUN']
---------------------------------------------
['ADP', 'PUNCT', 'NOUN', 'NOUN']
---------------------------------------------
['NOUN']
---------------------------------------------
['ADJ', 'NOUN', 'NOUN']
---------------------------------------------


# explaination of the code
So if a term repeats itself for more than once in a sequence we add a +

but in general regex terminology:
- we use `+` for one or more than one
- we use `*` for zero or more than one

So i am planning to use this if the current methods dont give results

In [116]:
# Cocvert a list of POS tag sequences into a regular exp pattern
def pos_to_regex(key_pos_list):    
    regex_patterns = []
    
    for pos_sequence in key_pos_list:
        regex_part = []
        prev_pos = ''
        pos_count = 0
        for pos in pos_sequence:
            if pos == prev_pos:
                pos_count += 1
            else:
                if prev_pos:
                    regex_part.append(prev_pos + ('+' if pos_count > 1 else ''))
                prev_pos = pos
                pos_count = 1
        regex_part.append(prev_pos + ('+' if pos_count > 1 else ''))        
        regex_patterns.append(','.join(regex_part))
    
    return regex_patterns


key_pos_list = [['ADJ', 'NOUN', 'NOUN'], ['ADJ', 'NOUN', 'NOUN', 'ADJ', 'ADJ'], ['ADJ', 'NOUN', 'NOUN', 'ADJ', 'ADJ', 'ADJ'], ['ADJ','ADJ']]
regex_patterns = pos_to_regex(key_pos_list)
print(regex_patterns)

['ADJ,NOUN+', 'ADJ,NOUN+,ADJ+', 'ADJ,NOUN+,ADJ+', 'ADJ+']


In [118]:
df['regex_pos'] = df['key_pos'].apply(pos_to_regex)

In [119]:
df['keyphrases'][0]

['NMR implementation',
 'quantum lattice gas algorithm',
 'quantum information processors',
 'fluid dynamics problems',
 'diffusion equation',
 'nonlinear Burgers equations',
 'nuclear magnetic resonance',
 'two-qubit quantum information.processors']

In [120]:
df['key_pos'][0]

[['PROPN', 'NOUN'],
 ['PROPN', 'PROPN', 'PROPN', 'PROPN'],
 ['PROPN', 'NOUN', 'NOUN'],
 ['ADJ', 'NOUN', 'NOUN'],
 ['NOUN', 'NOUN'],
 ['ADJ', 'PROPN', 'NOUN'],
 ['ADJ', 'ADJ', 'NOUN'],
 ['NUM', 'PUNCT', 'NOUN', 'NOUN', 'NOUN']]

In [121]:
df['regex_pos'][0]

['PROPN,NOUN',
 'PROPN+',
 'PROPN,NOUN+',
 'ADJ,NOUN+',
 'NOUN+',
 'ADJ,PROPN,NOUN',
 'ADJ+,NOUN',
 'NUM,PUNCT,NOUN+']

In [123]:
regex_pos_list = []
for i in df['regex_pos']:
    regex_pos_list += i

In [124]:
len(regex_pos_list)

9805

In [125]:
len(set(regex_pos_list))

668

We can see that i got down 9805 keywords down to 668 categories of classes out of which 

In [28]:
def list_to_sorted_dict(my_list):
    counts = Counter(my_list)
    sorted_dict = dict(sorted(counts.items(), key=lambda item: item[1], reverse=True))
    return sorted_dict

In [38]:
regex_dict = list_to_sorted_dict(regex_pos_list)
print(regex_dict)

{'NOUN+': 2157, 'ADJ,NOUN': 1551, 'NOUN': 854, 'ADJ,NOUN+': 764, 'PROPN,NOUN': 397, 'VERB,NOUN': 368, 'ADJ+,NOUN': 234, 'PROPN+': 222, 'PROPN': 206, 'VERB,NOUN+': 182, 'VERB': 153, 'ADJ,PUNCT,NOUN+': 134, 'NOUN,VERB': 129, 'PROPN,NOUN+': 96, 'PROPN+,NOUN': 95, 'ADJ+,NOUN+': 79, 'NOUN,PUNCT,NOUN+': 77, 'NOUN,PUNCT,VERB,NOUN': 74, 'VERB,ADJ,NOUN': 71, 'NOUN,VERB,NOUN': 51, 'NOUN,PUNCT,ADJ,NOUN': 44, 'NUM,NOUN': 42, 'NOUN,PUNCT,VERB,NOUN+': 42, 'ADJ,NOUN,VERB': 41, 'NOUN,ADP,NOUN': 40, 'NUM,PUNCT,NOUN+': 34, 'ADJ,PROPN,NOUN': 32, 'VERB,PUNCT,NOUN+': 30, 'PROPN,ADJ,NOUN': 28, 'ADJ,VERB,NOUN': 24, 'NOUN,PUNCT,NOUN': 24, 'NOUN,PROPN,NOUN': 22, 'ADV,VERB,NOUN': 20, 'VERB,PROPN,NOUN': 20, 'ADJ+,PUNCT,NOUN+': 20, 'ADJ,PUNCT,NOUN,ADJ,NOUN': 19, 'ADJ,PUNCT,ADJ,NOUN': 19, 'ADV,ADJ,NOUN': 19, 'NOUN+,VERB': 19, 'NOUN,ADJ,NOUN': 19, 'VERB,ADJ,NOUN+': 18, 'PROPN,VERB': 18, 'ADP,PUNCT,NOUN+': 17, 'NOUN,VERB,NOUN+': 17, 'PROPN,PUNCT,PROPN,NOUN': 17, 'NOUN,PROPN': 15, 'ADJ': 15, 'ADJ,PROPN': 14, 'X,NOUN+

In [188]:
import json

In [189]:
with open('probabilityRank.json', 'w') as json_file:
    json.dump(regex_dict, json_file, indent=4)

In [59]:
counts = 0 
for k,v in list(regex_dict.items())[:5]:
    counts += v
print(counts/len(regex_pos_list)*100)

58.36817950025497


- this says that the top 5 freq regex expression contribute to more than 58.36% of the total 

# n-gram model

In [43]:
df

id                                              title  \
0     761  Towards a NMR implementation of a quantum latt...   
1     724  Banking on SMA funds [separately managed accou...   
2    1371  Design methodology for diagnostic strategies f...   
3    1334                          A shy invariant of graphs   
4    1419  PacketVideo. One step ahead of the streaming w...   
..    ...                                                ...   
995  1124  Data extraction from the Web based on pre-defi...   
996   118       Sensorless control of induction motor drives   
997   635  Detection and estimation of abrupt changes in ...   
998  1260  A dataflow computer which accelerates executio...   
999  1225                       BT voices its support for IP   

                                              abstract  \
0    Recent theoretical results suggest that an arr...   
1    From investment management to technology to ba...   
2    This paper presents a method for the construct...   
3    Moving from a well known result of P.L. Hammer...   
4    Go beyond the hype, however, and it's clear th...   
..                                                 ...   
995  With the development of the Internet, the Worl...   
996  Controlled induction motor drives without mech...   
997  Detection of change-points in normal means is ...   
998  In the dataflow machine, it is important to av...   
999  BTexact's chief technology officer, Mick Reeve...   

                                            keyphrases  \
0    [NMR implementation, quantum lattice gas algor...   
1    [separately managed accounts, investment manag...   
2    [design methodology, diagnostic strategies, in...   
3    [graph invariant, induced odd cycles, minimum ...   
4    [PacketVideo, wireless devices, MPEG-4, wirele...   
..                                                 ...   
995  [data extraction, schema, Internet, informatio...   
996  [sensorless control, induction motor drives, r...   
997  [generalized likelihood ratio test statistic, ...   
998  [dataflow computer, sequential programs, prece...   
999  [BTexact, voice over DSL, virtual private netw...   

                                                  prmu  \
0                             [P, P, P, P, P, P, P, M]   
1                                   [P, P, P, P, P, P]   
2                                      [P, P, P, P, P]   
3                                   [P, P, P, P, P, P]   
4                             [P, P, P, P, P, P, P, R]   
..                                                 ...   
995                        [P, P, P, P, P, R, M, U, U]   
996  [P, P, P, P, P, P, P, P, P, P, P, P, P, P, P, ...   
997  [P, P, P, P, P, P, P, P, P, P, P, P, P, P, P, ...   
998               [P, P, P, P, P, P, P, P, P, R, M, M]   
999                                       [P, P, P, P]   

                                                tokens  \
0    [Recent, theoretical, results, suggest, that, ...   
1    [From, investment, management, to, technology,...   
2    [This, paper, presents, a, method, for, the, c...   
3    [Moving, from, a, well, known, result, of, P.L...   
4    [Go, beyond, the, hype, ,, however, ,, and, it...   
..                                                 ...   
995  [With, the, development, of, the, Internet, ,,...   
996  [Controlled, induction, motor, drives, without...   
997  [Detection, of, change, -, points, in, normal,...   
998  [In, the, dataflow, machine, ,, it, is, import...   
999  [BTexact, 's, chief, technology, officer, ,, M...   

                                                   pos  \
0    [ADJ, ADJ, NOUN, VERB, SCONJ, DET, NOUN, ADP, ...   
1    [ADP, NOUN, NOUN, ADP, NOUN, ADP, ADJ, PUNCT, ...   
2    [DET, NOUN, VERB, DET, NOUN, ADP, DET, NOUN, A...   
3    [VERB, ADP, DET, ADV, VERB, NOUN, ADP, PROPN, ...   
4    [VERB, ADP, DET, NOUN, PUNCT, ADV, PUNCT, CCON...   
..                                                 ...   
995  [ADP, DET, NOUN, ADP, DET, NOUN, PUNCT,

In [44]:
all_keypos_list = []
for i in df['key_pos']:
    all_keypos_list+=i

In [45]:
sorted(all_keypos_list, key=len, reverse=True)

[['PROPN',
  'PUNCT',
  'VERB',
  'PROPN',
  'PUNCT',
  'VERB',
  'NOUN',
  'PROPN',
  'NUM',
  'SYM',
  'PROPN',
  'PUNCT',
  'ADJ',
  'PUNCT',
  'NOUN',
  'NOUN',
  'NOUN'],
 ['ADJ',
  'PUNCT',
  'NOUN',
  'NUM',
  'PUNCT',
  'NOUN',
  'PUNCT',
  'ADP',
  'PUNCT',
  'NOUN',
  'PROPN',
  'SYM',
  'PROPN',
  'NOUN',
  'SYM',
  'NOUN'],
 ['NOUN',
  'PUNCT',
  'VERB',
  'NOUN',
  'NOUN',
  'SYM',
  'NOUN',
  'NOUN',
  'PUNCT',
  'ADP',
  'PUNCT',
  'NOUN',
  'NOUN',
  'NOUN'],
 ['VERB',
  'ADJ',
  'PUNCT',
  'ADJ',
  'PUNCT',
  'NOUN',
  'SYM',
  'ADJ',
  'PUNCT',
  'NOUN',
  'NOUN',
  'NOUN'],
 ['NOUN',
  'PUNCT',
  'NOUN',
  'PUNCT',
  'NOUN',
  'PUNCT',
  'NOUN',
  'PUNCT',
  'NOUN',
  'PUNCT',
  'NOUN',
  'NOUN'],
 ['NUM',
  'PUNCT',
  'NOUN',
  'NUM',
  'PUNCT',
  'NOUN',
  'ADJ',
  'PUNCT',
  'NOUN',
  'NOUN',
  'NOUN'],
 ['PROPN',
  'PUNCT',
  'VERB',
  'NOUN',
  'PUNCT',
  'VERB',
  'CCONJ',
  'NOUN',
  'PUNCT',
  'VERB',
  'NOUN'],
 ['VERB',
  'ADJ',
  'PUNCT',
  'NOUN',
  'SYM'

In [48]:
all_keywords_list = []
for i in df['key_pos']:
    all_keypos_list+=i

In [49]:
sorted(all_keywords_list, key=len, reverse=True)

[]

In [50]:
# labels = {'I am fine', 'He was born in 1995', 'The Eiffel tower', 'dogs'}
# preds = {'I am fine', 'born in 1995', 'Eiffel', 'dog'}

In [51]:
# len(labels.intersection(preds))

In [52]:
# def calculate_f1_score(actual_keyphrases, predicted_keyphrases):
#     """
#     Calculate the F1 score given actual and predicted keyphrases.

#     Parameters:
#     - actual_keyphrases: A list of strings, where each string is an actual keyphrase.
#     - predicted_keyphrases: A list of strings, where each string is a predicted keyphrase.

#     Returns:
#     - The F1 score as a float.
#     """
#     # Convert to sets for easier calculation
#     actual_set = set(actual_keyphrases)
#     predicted_set = set(predicted_keyphrases)

#     # Calculate precision and recall
#     true_positives = len(actual_set.intersection(predicted_set))
#     precision = true_positives / len(predicted_set) if predicted_set else 0
#     recall = true_positives / len(actual_set) if actual_set else 0

#     print(precision)
#     print(recall)
    
#     # Calculate F1 score
#     if precision + recall == 0:
#         f1_score = 0
#     else:
#         f1_score = 2 * (precision * recall) / (precision + recall)

#     return f1_score

# # Example usage
# actual_keyphrases = ['keyphrase one', 'keyphrase two', 'keyphrase three']
# predicted_keyphrases = ['keyphrase one', 'keyphrase four', 'keyphrase two']
# f1_score = calculate_f1_score(actual_keyphrases, predicted_keyphrases)
# print(f"F1 Score: {f1_score}")

In [53]:
# nouns_count = 0
# for i in df['pos']:
#     nouns_count += Counter(i)['NOUN']
# nouns_count

In [54]:
pred = ['Small banks',
 'attention',
 'investment management',
 'outsourcers',
 'technology',
 'back-office services',
 'SMA',
 'the SMA business',
 'their way',
 'the rewards']

In [56]:
true = df['keyphrases'][1]
true

['separately managed accounts',
 'investment management',
 'technology',
 'back-office services',
 'outsourcers',
 'small banks']

In [179]:
def categorize_and_analyze_keyphrases(key_pos_lists):
    categorized_keyphrases = {}
    for key_pos in key_pos_lists: 
        n_gram_length = len(key_pos)
        if categorized_keyphrases.get(n_gram_length):
            categorized_keyphrases[n_gram_length].append(key_pos)
        else:
            categorized_keyphrases[n_gram_length]=[key_pos]

    regex_freq_dict = {}
    for n_length, key_pos_list in categorized_keyphrases.items():
        regex_pattern_list = pos_to_regex(key_pos_list)
        regex_freq_dict[n_length] = list_to_sorted_dict(regex_pattern_list)

    regex_freq_dict = dict(sorted(regex_freq_dict.items(), key=lambda item: item[0], reverse=True))
    return categorized_keyphrases, regex_freq_dict

In [180]:
listt, n_gram_dict = categorize_and_analyze_keyphrases(key_pos)

In [181]:
listt.keys()

dict_keys([2, 4, 3, 5, 1, 9, 7, 11, 6, 8, 10, 12, 14, 17, 16])

In [190]:
n_gram_dict

{17: {'PROPN,PUNCT,VERB,PROPN,PUNCT,VERB,NOUN,PROPN,NUM,SYM,PROPN,PUNCT,ADJ,PUNCT,NOUN+': 1},
 16: {'ADJ,PUNCT,NOUN,NUM,PUNCT,NOUN,PUNCT,ADP,PUNCT,NOUN,PROPN,SYM,PROPN,NOUN,SYM,NOUN': 1},
 14: {'NOUN,PUNCT,VERB,NOUN+,SYM,NOUN+,PUNCT,ADP,PUNCT,NOUN+': 1},
 12: {'VERB,ADJ,PUNCT,ADJ,PUNCT,NOUN,SYM,ADJ,PUNCT,NOUN+': 1,
  'NOUN,PUNCT,NOUN,PUNCT,NOUN,PUNCT,NOUN,PUNCT,NOUN,PUNCT,NOUN+': 1},
 11: {'NUM,PUNCT,NOUN,NUM,PUNCT,NOUN,ADJ,PUNCT,NOUN+': 1,
  'PROPN,PUNCT,VERB,NOUN,PUNCT,VERB,CCONJ,NOUN,PUNCT,VERB,NOUN': 1,
  'VERB,ADJ,PUNCT,NOUN,SYM,ADJ+,PUNCT,NOUN+': 1,
  'NOUN,PUNCT,PROPN,PUNCT,PROPN,PUNCT,PROPN,PUNCT,PROPN,PUNCT,NOUN': 1},
 10: {'NUM,PUNCT,ADJ+,PUNCT,NOUN,ADJ,PUNCT,NOUN+': 1,
  'PROPN,PUNCT,VERB,NOUN,PUNCT,ADP,PUNCT,ADJ,NOUN+': 1,
  'ADV,PUNCT,VERB,NOUN,PUNCT,ADP,PUNCT,NOUN,VERB,NOUN': 1,
  'PROPN,ADJ,SYM,ADJ,NOUN,ADJ+,NOUN+': 1,
  'NOUN+,PART,ADJ,PUNCT,NOUN+,PROPN': 1,
  'PROPN,PUNCT,PROPN,NUM,CCONJ,PROPN,ADJ,NOUN,PROPN,NOUN': 1,
  'NOUN+,PUNCT,VERB,ADV,PUNCT,VERB,ADJ,NOUN+': 1,
 

In [191]:
with open('n_gram_dict.json', 'w') as json_file:
    json.dump(n_gram_dict, json_file, indent=4)